src/nemos/base_regressor.py

```diff
@@ -0,0 +1,418 @@
+            strength for proximal gradient methods.
+
+        prox:
+            Optional, the proximal projection operator.
+
+        *kwargs:
+            Keyword arguments for the jaxopt `solver.run` method.
+
+        Returns
+        -------
+        :
+            A tuple containing three callable functions:
+            - solver_init_state: Function to initialize the solver's state, necessary before starting the optimization.
+            - solver_update: Function to perform a single update step in the optimization process,
+            returning new parameters and state.
+            - solver_run: Function to execute the optimization process, applying multiple updates until a
+            stopping criterion is met.
+        """
+        # final check that solver is valid for chosen regularizer
+        if self.solver_name not in self.regularizer.allowed_solvers:
+            raise ValueError(
+                f"The solver: {self.solver_name} is not allowed for "
+                f"{self._regularizer.__class__} regularizaration. Allowed solvers are "
+                f"{self._regularizer.allowed_solvers}."
+            )
+
+        # only use penalized loss if not using proximal gradient descent
+        if self.solver_name != "ProximalGradient":
+            loss = self.regularizer.penalized_loss(
+                self._predict_and_compute_loss, self.regularizer_strength
+            )
+        else:
+            loss = self._predict_and_compute_loss
+
+        # check that the loss is Callable
+        utils.assert_is_callable(loss, "loss")
+
+        # some parsing to make sure solver gets instantiated properly
+        if self.solver_name == "ProximalGradient":
+            self.solver_kwargs.update(prox=self.regularizer.get_proximal_operator())
+            # add self.regularizer_strength to args
+            args += (self.regularizer_strength,)
+        if self.solver_name == "LBFGSB":
+            if "bounds" not in kwargs.keys():
+                warnings.warn(
+                    "Bounds must be provided for LBFGSB. Reverting back to LBFGS solver."
+                )
+            self.solver_name = "LBFGS"
+
+        # instantiate the solver
+        solver = getattr(jaxopt, self._solver_name)(fun=loss, **self.solver_kwargs)
+
+        def solver_run(
+            init_params: Tuple[DESIGN_INPUT_TYPE, jnp.ndarray], *run_args: jnp.ndarray
+        ) -> jaxopt.OptStep:
+            return solver.run(init_params, *args, *run_args, **kwargs)
+
+        def solver_update(params, state, *run_args, **run_kwargs) -> jaxopt.OptStep:
+            return solver.update(
+                params, state, *args, *run_args, **kwargs, **run_kwargs
+            )
+
+        def solver_init_state(params, state, *run_args, **run_kwargs) -> NamedTuple:
+            return solver.init_state(
+                params, state, *args, *run_args, **kwargs, **run_kwargs
+            )
+
+        return solver_init_state, solver_update, solver_run
```

Comment:
out of curiosity: why return these instead of just setting them as attributes here?

Reply:
I think setting the attribute makes sense. It is a leftover from when we had this function as part of the `Regularizer`

Reply:
We just didn't move those properties from the `GLM` class to the `BaseRegressor`, but we could if we wanted to I think. 

@BalzaniEdoardo thoughts on doing this?

Reply:
Yeah, I think this should be a common attribute of any BaseRegressor

Reply:
now setting attributes here.